src/flint/fmpz_mod_mat.jl

```diff
@@ -0,0 +1,863 @@
+   fl && return false
+   fl = (ncols(x) != ncols(y)) && (nrows(x) != nrows(y))
+   fl && throw && error("Matrices have wrong dimensions")
+   return !fl
+end
+
+###############################################################################
+#
+#   Similar
+#
+###############################################################################
+
+function similar(::MatElem, R::FmpzModRing, r::Int, c::Int)
+   z = fmpz_mod_mat(r, c, R.n)
+   z.base_ring = R
+   return z
+end
+
+################################################################################
+#
+#  Manipulation
+#
+################################################################################
+
+@inline function getindex(a::T, i::Int, j::Int) where T <: Zmod_fmpz_mat
+  @boundscheck Generic._checkbounds(a, i, j)
+  u = fmpz()
+  ccall((:fmpz_mod_mat_get_entry, :libflint), Nothing,
+              (Ref{fmpz}, Ref{T}, Int, Int), u, a, i - 1 , j - 1)
+  return fmpz_mod(u, base_ring(a)) # no reduction needed
+end
+
+# as above, but as a plain fmpz, no bounds checking
+function getindex_raw(a::T, i::Int, j::Int) where T <: Zmod_fmpz_mat
+  u = fmpz()
+  ccall((:fmpz_mod_mat_get_entry, :libflint), Nothing,
+                 (Ref{fmpz}, Ref{T}, Int, Int), u, a, i - 1, j - 1)
+  return u
+end
+
+@inline function setindex!(a::T, u::fmpz, i::Int, j::Int) where T <: Zmod_fmpz_mat
+  @boundscheck Generic._checkbounds(a, i, j)
+  R = base_ring(a)
+  setindex_raw!(a, mod(u, R.n), i, j)
+end
+
+@inline function setindex!(a::T, u::fmpz_mod, i::Int, j::Int) where T <: Zmod_fmpz_mat
+  @boundscheck Generic._checkbounds(a, i, j)
+  (base_ring(a) != parent(u)) && error("Parent objects must coincide")
+  setindex_raw!(a, u.data, i, j) # no reduction needed
+end
+
+function setindex!(a::T, u::Integer, i::Int, j::Int) where T <: Zmod_fmpz_mat
+   setindex!(a, fmpz(u), i, j)
+end
+
+# as per setindex! but no reduction mod n and no bounds checking
+@inline function setindex_raw!(a::T, u::fmpz, i::Int, j::Int) where T <: Zmod_fmpz_mat
+  ccall((:fmpz_mod_mat_set_entry, :libflint), Nothing,
+        (Ref{T}, Int, Int, Ref{fmpz}), a, i - 1, j - 1, u)
+end
+
+function deepcopy_internal(a::fmpz_mod_mat, dict::IdDict)
+  z = fmpz_mod_mat(nrows(a), ncols(a), parent(a).n)
+  if isdefined(a, :base_ring)
+    z.base_ring = a.base_ring
+  end
+  ccall((:fmpz_mod_mat_set, :libflint), Nothing,
+          (Ref{fmpz_mod_mat}, Ref{fmpz_mod_mat}), z, a)
+  return z
+end
+
+nrows(a::T) where T <: Zmod_fmpz_mat = a.r
+
+ncols(a::T) where T <: Zmod_fmpz_mat = a.c
+
+nrows(a::FmpzModMatSpace) = a.nrows
+
+ncols(a::FmpzModMatSpace) = a.ncols
+
+function parent(a::T, cached::Bool = true) where T <: Zmod_fmpz_mat
+   MatrixSpace(base_ring(a), nrows(a), ncols(a), cached)
+end
+
+base_ring(a::FmpzModMatSpace) = a.base_ring
+
+base_ring(a::T) where T <: Zmod_fmpz_mat = a.base_ring
+
+zero(a::FmpzModMatSpace) = a()
+
+function one(a::FmpzModMatSpace)
+  (nrows(a) != ncols(a)) && error("Matrices must be quadratic")
```

Comment:
I was not aware that "quadratic matrix" means "square matrix", "square" might be preferable for non-native speakers.

Reply:
It doesn't mean that. This is something that German people write. If I put it there, it is a copy-paste error. I'll fix it.

Reply:
Fixed.